File: admin/src/services/councilMemberService.ts
import api from "./api";

export interface CouncilMember {
  id: number;
  name: string;
  party: string;
  image: string | null;
  created_at: string;
  updated_at: string;
}

export interface CouncilMemberCreateData {
  name: string;
  party: string;
  image: string | null;
}

export interface CouncilMemberUpdateData {
  name: string;
  party: string;
  image: string | null;
}

export const getCouncilMembers = async (): Promise<CouncilMember[]> => {
  const response = await api.get("/council-members/");
  return response.data;
};

export const getCouncilMember = async (
  id: number
): Promise<CouncilMember> => {
  const response = await api.get(`/council-members/${id}`);
  return response.data;
};

export const createCouncilMember = async (
  data: CouncilMemberCreateData
): Promise<CouncilMember> => {
  const response = await api.post("/council-members/", data);
  return response.data;
};

export const updateCouncilMember = async (
  id: number,
  data: CouncilMemberUpdateData
): Promise<CouncilMember> => {
  const response = await api.put(`/council-members/${id}`, data);
  return response.data;
};

export const deleteCouncilMember = async (
  id: number
): Promise<void> => {
  await api.delete(`/council-members/${id}`);
};

File: backend/app/services/council_member_service.py
from datetime import datetime

from sqlalchemy.orm import Session

from app.models.council_member import CouncilMember
from app.schemas.council_member import (
    CouncilMemberCreate,
    CouncilMemberUpdate,
)


def get_all_council_members(
    db: Session,
) -> list[CouncilMember]:
    return (
        db.query(CouncilMember)
        .order_by(CouncilMember.id.asc())
        .all()
    )


def get_council_member(
    db: Session,
    council_member_id: int,
) -> CouncilMember | None:
    return (
        db.query(CouncilMember)
        .filter(
            CouncilMember.id == council_member_id
        )
        .first()
    )


def create_council_member(
    db: Session,
    data: CouncilMemberCreate,
) -> CouncilMember:
    council_member = CouncilMember(
        name=data.name,
        party=data.party,
        image=data.image,
    )

    db.add(council_member)
    db.commit()
    db.refresh(council_member)

    return council_member


def update_council_member(
    db: Session,
    council_member_id: int,
    data: CouncilMemberUpdate,
) -> CouncilMember | None:
    council_member = get_council_member(
        db,
        council_member_id,
    )

    if not council_member:
        return None

    council_member.name = data.name
    council_member.party = data.party
    council_member.image = data.image
    council_member.updated_at = datetime.utcnow()

    db.commit()
    db.refresh(council_member)

    return council_member


def delete_council_member(
    db: Session,
    council_member_id: int,
) -> bool:
    council_member = get_council_member(
        db,
        council_member_id,
    )

    if not council_member:
        return False

    db.delete(council_member)
    db.commit()

    return True